homeassistant/components/xiaomi_miio/humidifier.py

```diff
@@ -0,0 +1,387 @@
+from homeassistant.util.percentage import percentage_to_ranged_value
+
+from .const import (
+    CONF_DEVICE,
+    CONF_FLOW_TYPE,
+    DOMAIN,
+    KEY_COORDINATOR,
+    KEY_DEVICE,
+    MODEL_AIRHUMIDIFIER_CA1,
+    MODEL_AIRHUMIDIFIER_CA4,
+    MODEL_AIRHUMIDIFIER_CB1,
+    MODELS_HUMIDIFIER_MIOT,
+    SERVICE_SET_TARGET_HUMIDITY,
+)
+from .device import XiaomiCoordinatedMiioEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "Xiaomi Miio Device"
+DATA_KEY = "fan.xiaomi_miio"
+
+CONF_MODEL = "model"
+
+# Air Humidifier
+ATTR_TARGET_HUMIDITY = "target_humidity"
+ATTR_TRANS_LEVEL = "trans_level"
+ATTR_HARDWARE_VERSION = "hardware_version"
+
+SERVICE_TO_METHOD = {
+    SERVICE_SET_TARGET_HUMIDITY: {
+        "method": "async_set_target_humidity",
+    },
+}
+
+AVAILABLE_ATTRIBUTES = {
+    ATTR_MODE: "mode",
+    ATTR_TARGET_HUMIDITY: "target_humidity",
+}
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Humidifier from a config entry."""
+    entities = []
+
+    if not config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
+        return
+
+    if DATA_KEY not in hass.data:
+        hass.data[DATA_KEY] = {}
+
+    host = config_entry.data[CONF_HOST]
+    token = config_entry.data[CONF_TOKEN]
+    name = config_entry.title
+    model = config_entry.data[CONF_MODEL]
+    unique_id = config_entry.unique_id
+
+    _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
+
+    if model in MODELS_HUMIDIFIER_MIOT:
+        air_humidifier = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+        entity = XiaomiAirHumidifierMiot(
+            name,
+            air_humidifier,
+            config_entry,
+            unique_id,
+            hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR],
+        )
+    else:
+        air_humidifier = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+        entity = XiaomiAirHumidifier(
+            name,
+            air_humidifier,
+            config_entry,
+            unique_id,
+            hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR],
+        )
+
+    hass.data[DATA_KEY][host] = entity
+    entities.append(entity)
+
+    async_add_entities(entities, update_before_add=True)
+
+
+class XiaomiGenericHumidifier(XiaomiCoordinatedMiioEntity, HumidifierEntity):
+    """Representation of a generic Xiaomi humidifier device."""
+
+    _attr_device_class = DEVICE_CLASS_HUMIDIFIER
+    _attr_supported_features = SUPPORT_MODES
+    _attr_should_poll = True
+
+    def __init__(self, name, device, entry, unique_id, coordinator):
+        """Initialize the generic Xiaomi device."""
+        super().__init__(name, device, entry, unique_id, coordinator=coordinator)
+
+        self._available = False
+        self._state = None
+        self._attributes = {}
+        self._skip_update = False
+        self._available_modes = []
+        self._mode = None
+        self._available_attributes = AVAILABLE_ATTRIBUTES
```

Comment:
We can use the constant directly. We don't need the instance attribute.

Reply:
Done